safe/reportv4/impact_report.py

```diff
@@ -0,0 +1,770 @@
+            except KeywordNotFoundError:
+                pass
+            except Exception:  # pylint: disable=broad-except
+                pass
+        return legend_attribute_dict
+
+    def process_component(self):
+        """Process context for each component and a given template"""
+        for component in self.metadata.components:
+            # load extractors
+            if callable(component.extractor):
+                _extractor_method = component.extractor
+            else:
+                _package_name = '%s.extractors.%s' % (
+                    self.metadata.key,
+                    component.key
+                )
+                # replace dash with underscores
+                _package_name = _package_name.replace('-', '_')
+                _extractor_path = os.path.join(
+                    self.metadata.template_folder,
+                    component.extractor
+                )
+                _module = imp.load_source(_package_name, _extractor_path)
+                _extractor_method = getattr(_module, 'extractor')
+            try:
+                # method signature:
+                #  - this ImpactReport
+                #  - this component
+                context = _extractor_method(self, component)
+                component.context = context
+            except Exception as exc:
+                LOGGER.exception(exc)
+
+            # load processor
+            if callable(component.processor):
+                _renderer = component.processor
+            else:
+                _package_name = '%s.renderer.%s' % (
+                    self.metadata.key,
+                    component.key
+                )
+                # replace dash with underscores
+                _package_name = _package_name.replace('-', '_')
+                _renderer_path = os.path.join(
+                    self.metadata.template_folder,
+                    component.processor
+                )
+                _module = imp.load_source(_package_name, _renderer_path)
+                _renderer = getattr(_module, 'renderer')
+            try:
+                # method signature:
+                output = _renderer(self, component)
+                component.output = output
+            except Exception as exc:
+                LOGGER.exception(exc)
+
+    def setup_composition(self):
+        """Set up the composition ready."""
+        # noinspection PyUnresolvedReferences
+        self._template_composition.composition.setPlotStyle(
+            QgsComposition.Preview)
+        self._template_composition.composition.setPrintResolution(
+            self.page_dpi)
+        self._template_composition.composition.setPrintAsRaster(True)
+
+    def load_template(self):
+        """Load the template to composition."""
+        # Get information for substitutions
+        # date, time and plugin version
+        date_time = self._keyword_io.read_keywords(self.layer, 'time_stamp')
```

Comment:
Are you sure you need to use keyword_io ? We have now the monkey patching for keywords.